allow external backend codegen to toggle whether to generate out= and inplace kernels

bdhirsh opened the pull request:
The external codegen pipeline currently uses a `use_out_as_primary` flag to determine if we should auto-generate `inplace` and `out=` kernels given an `out-of-place` kernel from the backend. This is only something that probably makes sense for graph-mode compilers like XLA/LTC, and is currently hardcoded to support XLA.

This PR defaults it to the XLA setting, but exposes it as a toggle in the yaml. Intel will want this.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69686
* #68531
* __->__ #68530
* #68529

Differential Revision: [D32498569](https://our.internmc.facebook.com/intern/diff/D32498569)

bdhirsh committed:
allow external backend codegen to toggle whether to generate out= and inplace kernels

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498569).

bdhirsh commented on test/test_gen_backend_stubs.py:
```diff
@@ -13,6 +14,11 @@
 
+    def setUp(self) -> None:
+        global _GLOBAL_PARSE_NATIVE_YAML_CACHE
+        _GLOBAL_PARSE_NATIVE_YAML_CACHE.clear()
```
@albanD I realized that this test hasn't been running in CI, and it's been failing because of this cache.

I think this came up in a discussion a while back with Ed - the easiest thing for me to do to fix the test is just to clear the cache between runs. I checked and the test takes ~5-10 seconds to run, which isn't too bad.

albanD replied:
Would using functools.cache avoid this problem?

bdhirsh replied:
I don't think so - I quickly tried replacing `_GLOBAL_PARSE_NATIVE_YAML_CACHE with a `functools.cache` decorator and still see the failures.

albanD reviewed (approved):
LGTM

guoyejun commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -78,7 +83,7 @@ def create_backend_index(backend_ops: List[str], dispatch_key: DispatchKey) -> B
 
-        backend_idx = create_backend_index(supported, backend_key)
+        backend_idx = create_backend_index(supported, backend_key, use_out_as_primary=use_out_as_primary)
         assert backend_key not in backend_indices
```
one open:
i use pudb to debug python program, i usually restart the program within pudb, but it hits the assert here in the second run. The reason is that _GLOBAL_PARSE_NATIVE_YAML_CACHE is not clear in the beginning of the second run, and so the backend_key is already in backend_indices. Any comment? thanks.

bdhirsh replied:
cc @albanD. We added the cache because it gives a pretty large speed-up to the build, but this is probably a downside. (Although hopefully running the backend codegen multiple times in the python runtime isn't something we have to do a lot). You can probably either explicitly clear the cache in the pudb session (?), or just restart pudb.

albanD replied:
We do have the assumption that the codegen is a fresh python run everytime yes. 
You can get away with not restarting by clearing the cache etc (as we do in some testing) but that is brittle at best.

guoyejun replied:
thanks both.

bdhirsh committed:
Update on "allow external backend codegen to toggle whether to generate out= and inplace kernels"

The external codegen pipeline currently uses a `use_out_as_primary` flag to determine if we should auto-generate `inplace` and `out=` kernels given an `out-of-place` kernel from the backend. This is only something that probably makes sense for graph-mode compilers like XLA/LTC, and is currently hardcoded to support XLA.

This PR defaults it to the XLA setting, but exposes it as a toggle in the yaml. Intel will want this.


Differential Revision: [D32498569](https://our.internmc.facebook.com/intern/diff/D32498569)

[ghstack-poisoned]

bdhirsh commented on test/test_gen_backend_stubs.py:
```diff
@@ -2,13 +2,22 @@
+    spec.loader.exec_module(module)
+    return module
+
+gen_backend_stubs = import_codegen('gen_backend_stubs', 'tools/codegen/gen_backend_stubs.py')
```
@albanD I think this has come up before - right now CI isn't happy with this test because the tools directory isn't part of `sys.path` in CI. This is the best I could come up with as a workaround without modifying sys.path: I'm using `importlib` to manually import a file, given its full path.

albanD replied:
As discussed offline, this can move to the test folder inside `tools/` that run in a different CI job

bdhirsh commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -25,6 +25,7 @@ def parse_backend_yaml(
         grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
         backend_indices: Dict[DispatchKey, BackendIndex]
 ) -> ParsedExternalYaml:
+    _GLOBAL_PARSE_NATIVE_YAML_CACHE.clear()
```
One consequence of https://github.com/pytorch/pytorch/pull/68530/files#r754702093 is that I can't modify this cache directly in `test_gen_backend_stubs.py` - I think it shows up as a separate variable when I use the `importlib` machinery. It seems... okay to clear the cache here, because at least right now, you'll never want to be in a situation where you want to call `parse_backend_yaml()` multiple times on the same cached yaml output.

bdhirsh committed:
Update on "allow external backend codegen to toggle whether to generate out= and inplace kernels"

The external codegen pipeline currently uses a `use_out_as_primary` flag to determine if we should auto-generate `inplace` and `out=` kernels given an `out-of-place` kernel from the backend. This is only something that probably makes sense for graph-mode compilers like XLA/LTC, and is currently hardcoded to support XLA.

This PR defaults it to the XLA setting, but exposes it as a toggle in the yaml. Intel will want this.


Differential Revision: [D32498569](https://our.internmc.facebook.com/intern/diff/D32498569)

[ghstack-poisoned]

bdhirsh committed:
Update on "allow external backend codegen to toggle whether to generate out= and inplace kernels"

The external codegen pipeline currently uses a `use_out_as_primary` flag to determine if we should auto-generate `inplace` and `out=` kernels given an `out-of-place` kernel from the backend. This is only something that probably makes sense for graph-mode compilers like XLA/LTC, and is currently hardcoded to support XLA.

This PR defaults it to the XLA setting, but exposes it as a toggle in the yaml. Intel will want this.


Differential Revision: [D32498569](https://our.internmc.facebook.com/intern/diff/D32498569)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498569).

bdhirsh committed:
Update on "allow external backend codegen to toggle whether to generate out= and inplace kernels"

The external codegen pipeline currently uses a `use_out_as_primary` flag to determine if we should auto-generate `inplace` and `out=` kernels given an `out-of-place` kernel from the backend. This is only something that probably makes sense for graph-mode compilers like XLA/LTC, and is currently hardcoded to support XLA.

This PR defaults it to the XLA setting, but exposes it as a toggle in the yaml. Intel will want this.


Differential Revision: [D32498569](https://our.internmc.facebook.com/intern/diff/D32498569)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498569).

albanD reviewed (approved):
LGTM

bdhirsh committed:
Update on "allow external backend codegen to toggle whether to generate out= and inplace kernels"

The external codegen pipeline currently uses a `use_out_as_primary` flag to determine if we should auto-generate `inplace` and `out=` kernels given an `out-of-place` kernel from the backend. This is only something that probably makes sense for graph-mode compilers like XLA/LTC, and is currently hardcoded to support XLA.

This PR defaults it to the XLA setting, but exposes it as a toggle in the yaml. Intel will want this.


Differential Revision: [D32498569](https://our.internmc.facebook.com/intern/diff/D32498569)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498569).
